### DISCUSSION TITLE
Provide native lib for libsignal with GraalVM

### DISCUSSION THREAD

**Discussion by RichardFevrier** [AsamK/signal-cli#689] [Q&A]
Hi everyone 👋 thank you for your amazing work!

I'm facing an issue when I try to include `signal-client-java-0.2.3.jar` & `zkgroup-java-0.7.0.jar` modified using compiled versions of `libsignal-client-java-0.2.3` & `zkgroup-0.7.2` AND GraalVM native image for aarch64 (Raspberry Pi 4 thru a docker image).

If I use the "usual" way without GraalVM everything works like a charm!.. except that starting a JVM every time we need to send a message is a no-go to me in terms of performance.

So I decided to try to include GraalVM in the process of building this library;
It works! But not for the modified versions of `signal-client-java-0.2.3.jar` & `zkgroup-java-0.7.0.jar`.

Here is what I've done so far (sorry if it doesn't make sense, I absolutely don't know how Gradle, GraalVM.. works):

- Fetching & compiling (sources) `libsignal-client-java-0.2.3` & `zkgroup-0.7.2`
- Fetching (sources) `signal-cli-0.8.1`
- Compiling `signal-cli-0.8.1` with `gradle build && gradle installDist` BUT not `gradle distTar`... which doesn't seem related to GraalVM
- Patching (zip) `signal-client-java-0.2.3.jar` & `zkgroup-java-0.7.0.jar` with the results of `libsignal-client-java-0.2.3` & `zkgroup-0.7.2` compilations
- Finally compiling `signal-cli-0.8.1` with `gradle assembleNativeImage`

Unfortunately `gradle assembleNativeImage` doesn't seem to use the result of `gradle build && gradle installDist` to perform his tasks..

I hope @AsamK or/and @michaelkebe could take a look into this issue.

I'm ready to perform tests if you need me to.

Thanks a lot for your time.
Best regards.

**Comment by RichardFevrier** [comment]
Some news here, I read more about GraalVM and I was able to get it working.

Writing here if it can help someone else: 

- Fetching & compiling (sources) `libsignal-client-java-0.2.3` & `zkgroup-0.7.2`
- Fetching (sources) `signal-cli-0.8.1`
- Copying the results of `libsignal-client-java-0.2.3` & `zkgroup-0.7.2` compilations (`libsignal_jni.so` & `libzkgroup.so`) into `signal-cli-0.8.1/lib/src/main/resources`
- Finally compiling `signal-cli-0.8.1` with `gradle assembleNativeImage`

One thing not working tho is the `support for new group v2` because apparently `libzkgroup` is missing..
I will try to investigate further.

Anyway, performance is great now!
Maybe it would be a good idea to add a little bit more documentation.

Thanks a lot for your time.
Best regards.

**Comment by mericon** [comment]
I Compiled as you described and Got an native build but IT says that lib misssing required native library dependency: libsignal-client. How can i fix that?

Got IT Workin with v0.8.1

**Comment by RichardFevrier** [comment]
@mericon do you have the `support for new group v2` missing too ?
